python/ray/serve/_private/router.py

```diff
@@ -594,15 +611,19 @@ async def choose_two_replicas_with_backoff(
                     )
             elif (
                 self._prefer_local_routing
-                and backoff_index == 0
-                and len(self._colocated_replica_ids) > 0
+                and not tried_same_node
+                and len(self._node_colocated_replica_ids) > 0
             ):
-                # Attempt to schedule requests to a replica on the same node in the
-                # first iteration only.
-                candidate_replica_ids = self._colocated_replica_ids
+                # Attempt to schedule requests to replicas on the same node at most once
+                candidate_replica_ids = self._node_colocated_replica_ids
+                tried_same_node = True
+            elif not tried_same_az and len(self._az_colocated_replica_ids) > 0:
```

Comment:
I think we should also add a feature flag to be able to switch back to old behavior (just in case)

Reply:
Not opposed cc: @akshay-anyscale

Reply:
I added a feature flag which is fetched in `Router`. It is not exposed as an option in handle.